### PR TITLE
sys: refactor synchronous uart

### PR DESCRIPTION
**Pull request by kaspar030**
uart_stdio used to hard-code it's method of passing incoming uart bytes to "userspace".

This PR adds a new module (isrpipe) which does exactly that, making it possible to re-use that functionality.
The second commit makes uart_stdio use the new module.

On SAMR21, this adds ~30 bytes of code, unless compiled with LTO, which compiles to the same size.

### CODE REVIEW

**Comment by haukepetersen** [comment]
Looks good to me (though I have not tested it yet).

Looks awefully familiar to some PR that @Kijewski did a long long time ago :-)

Wiil test ASAP.

File: sys/isrpipe/isrpipe.c [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    mutex_unlock(&isrpipe->mutex);
+
+    return res;
+}
+
+int isrpipe_read(isrpipe_t *isrpipe, char *buffer, size_t count)
+{
+    int res;
+
+    while (!(res = tsrb_get(&isrpipe->tsrb, buffer, count))) {
```
Comment (lebrush) [inline sys/isrpipe/isrpipe.c]: I had implemented those methods myself in the past and I faced one issue with a [very] slow uart device. 
The code for `tsrb_get` is:

```C
int tsrb_get(tsrb_t *rb, char *dst, size_t n)
{
    size_t tmp = n;
    while (tmp && !tsrb_empty(rb)) {
        *dst++ = _pop(rb);
        tmp--;
    }
    return (n - tmp);
}
```

Imagine you want to read 5 chars but there's only 1 available at a time. 
First time you call `tsrb_get` the `rb` is empty so you get 0 and you lock. A single character arrives, the mutex is unlocked and you call `tsrb_get` again: now it returns 1, so you don't lock anymore and you return 1 single byte. Also could happen that there's a leftover byte in the uart when you call the read method: same thing.
Is this the expected behaviour? 

Maybe this could be tackled by defining two different methods: one would have this behaviour and one would block until 5 bytes were really available. Or maybe the documentation updated. Or am I missing something...?

Comment (kaspar030) [inline sys/isrpipe/isrpipe.c]: > Looks awefully familiar to some PR that @Kijewski did a long long time ago :-)

You mean the pipe module?

**Comment by kaspar030** [comment]
> Looks awefully familiar to some PR that @Kijewski did a long long time ago :-)

You mean the pipe module?

File: sys/isrpipe/isrpipe.c (hunk 1 above)
Comment (kaspar030) [inline sys/isrpipe/isrpipe.c]: Currently, this is expected. isrpipe_read() returns as soon as there is data available, which with uarts, is usually just one initially.

I have "read with timeout" and "read all with timeout" functions in the queue, but wanted to keep the PR small for now.

**Comment by haukepetersen** [comment]
Yupp, but maybe it was just the name that came back to mind, don't remember exactly what it was about...

File: sys/isrpipe/isrpipe.c (hunk 1 above)
Comment (haukepetersen) [inline sys/isrpipe/isrpipe.c]: @lebrush: I think for the context of stdio_uart that 'single byte read' behavior is intended, as we never know how many byte we actually expect... But for other use-cases I think this could be improved?!

@kaspar030: have you spend thought also on on functions for `return once N byte are available` as described above, and on `return once last received byte equals SOMEVAL`? The latter could be quite useful for the stdio implementation waiting for `\n` chars...

Comment (kaspar030) [inline sys/isrpipe/isrpipe.c]: > I have "read with timeout" and "read all with timeout" functions in the queue, but wanted to keep the PR small for now.

Comment (kaspar030) [inline sys/isrpipe/isrpipe.c]: > return once last received byte equals SOMEVAL

Shouldn't that be a generic readline function? (handle \r\n -> \n etc)

Comment (haukepetersen) [inline sys/isrpipe/isrpipe.c]: I guess we could go either way, but for the shell case, `\n` would work for both, and we could also specify, that you must use Unix line endings when communicating with the RIOT shell...

Anyway, not really part of this PR...

Comment (kaspar030) [inline sys/isrpipe/isrpipe.c]: > for the shell case, \n would work for both

Not really, because then you'd have, depending on the terminal, the line with or without trailing ```\r```. But yeah, out of scope. :)

File: sys/isrpipe/isrpipe.c [hunk 2]
```diff
@@ -0,0 +1,85 @@
+#include <errno.h>
```
Comment (haukepetersen) [inline sys/isrpipe/isrpipe.c]: doxygen/license header missing...

**Comment by haukepetersen** [comment]
and it seems some cppcheck issues as well...

**Comment by kaspar030** [comment]
- fixed static-tests complaints
- rebased

**Comment by haukepetersen** [comment]
ACK and go
